ENH: -d^. to point to the top of the current dataset

yarikoptic opened the pull request:
It is often desired to reference current dataset. E.g. to add an existing
dataset as subdataset to the current dataset somewhere not at the top of the
dataset directories tree. Especially in quite nested hierarchies it
becomes a burden to figure out where is the boundary of the dataset to specify
it in -d option.  -d .  would point to the (not installed, and possibly illegit
see #3211) subdataset in current directory, whenever actual need is to specify
smth like -d ../../..  .

With -d^. it becomes possible to say "refer to the current dataset I am in".

The only gotcha I am aware of is that shell might like to use "^" as
a quick substitution for previous command, that is why I specified it
as -d^., i.e. without space - then seems to work as intended -- running
"datalad add -d^. ./subds"  would add subds in current subdirectory to the
dataset without requiring full path in -d.

Depending on the outcome of #3230 discussion, this change might not be really
needed, e.g. if default context for operations (including on subdatasets) would
migrate to be of "current dataset".  Otherwise, if -d would be needed to "bind"
the context to current dataset - at least with this helper it would be easy to
unambigously reference current dataset regardless of the position in the
directories hierarchy.

If you have a better suggestion instead of `^.` - please suggest.  My choice was because we have `^` to point to supermost dataset, thus use it to point to the (very) top. So it kinda made sense to me to point to the top of dataset curdir belongs to with `^.`, and have it on the left because that target path is to the left of curdir (although `.^` probably would have been better for use in shell due to aforementioned gotcha)

### Instructions

- [x] **Delete these instructions**. :-)

mih commented:
FTR: I think we should be reluctant to add general meaning to argument values that are only needed to shoehorn particular behavior into the existing machinery. This issue is a flavor of "the shell is to dumb". In Python code this should be rather unnecessary (one would use proper Dataset instances). Maybe it is worth implementing such argument transformation in the initial cmdline API layer, and not further complicate path interpretation globally.

Given that the example uses `add`, I would prefer to discuss this in the context of `rev-save` which doesn't have many of add's issues and also consider my proposal in https://github.com/datalad/datalad/issues/3230#issuecomment-473661218 for any command to stop behaving differently with or without `-d` given -- which would make the entire issue go away, without adding more decision making logic.

yarikoptic commented:
> Maybe it is worth implementing such argument transformation in the initial cmdline API layer

It could be done there but I do not see a real need to break concurrency between python and cmdline APIs for this.
Re #3230 - indeed, if we make commands behave similar with and without `-d` such shortcut might not be needed (as I noted in the PR description), so I am ok for this PR to linger around until we have a clear(er) picture.

kyleam commented:
It makes sense to see how things settle, but 

> > Maybe it is worth implementing such argument transformation in the initial cmdline API layer
> 
> It could be done there but I do not see a real need to break concurrency between python and cmdline APIs for this.

... from the standpoint of someone inspecting the code, I like that in this patch `^.` is handled in the same location as `///`.

driusan commented:
I think the usage of "^" is likely to cause some confusion, since it already has meaning to git and the usage here is context sensitive. With this change places referring to datasets it means one thing, while places specifying git revisions it means something completely unrelated.

mih commented:
Ok, I am with @kyleam now. Good to go.

It would be very good, though, to have these symbols and behavior documented somewhere. Thanks in advance!

mih commented:
What shall we do with this one?

mih commented:
Ping!

mih commented:
Shall I change the base and merge this one into master?

yarikoptic commented:
Without agreeing on smth like 1. of https://github.com/datalad/datalad/issues/3905 I guess we should indeed have something like this. I have rebased and added minimal doc mentioning.  Note: Since `--dataset` is command specific, duplicated across all the commands, we do not mention `///` or `^` anywere "in the API docs", so the same remains for `^.`

kyleam commented:
@yarikoptic Thanks for the updates.
